tf_agents/agents/behavioral_cloning/behavioral_cloning_agent.py

```diff
@@ -256,7 +256,7 @@ def _continuous_loss_fn(self, experience, training: bool = False):
     else:
       bc_action = bc_output
 
-    losses = tf.nest.map_structure(tf.losses.mse, experience.action, bc_action)
+    losses = tf.nest.map_structure(tf.math.squared_difference, experience.action, bc_action)
```

Comment:
Can you calculate the mean here?  changing from mse to squared_difference means this loss is no longer normalized by  batch size, and means existing learning rate may no longer work the same waay.

Reply:
I'm not sure that's necessary.

I think the return of this function needs to be per-example loss, and batch-size scaling is handled by [`common.aggregate_losses`](https://www.tensorflow.org/agents/api_docs/python/tf_agents/utils/common/aggregate_losses)

Which is called here right afterwards: https://github.com/tensorflow/agents/blob/e1d637f224c225491c7424e8c4b6fa21f334f36e/tf_agents/agents/behavioral_cloning/behavioral_cloning_agent.py#L295

The previous mean calculation was redundant.

Reply:
Ah you're partly right; the aggregation is not over batch but over *time*.  So you're replacing average loss over time w/ sum loss over time; and that means if you have an RNN nework and subsequences, if you change the subsequence lengths then the loss will change.  I still think we want the mean here.

Reply:
MSE is not dividing by per_example_loss.shape[0].  It normalizes by the count of non-batch entries:

```python
 tf.losses.mse(tf.constant([[0.5, 0.5, 0.5]]), tf.constant([[1.0, 1.0, 1.0]]))
>>> numpy=array([0.25], dtype=float32)
tf.losses.mse(tf.constant(2 * [[0.5, 0.5, 0.5]]), tf.constant(2 * [[1.0, 1.0, 1.0]]))
>>> numpy=array([0.25, 0.25], dtype=float32)
```

Reply:
Ah, there we go, my mistake, I figured out what was happening on my end to cause the warning.

`tf.losses.mse` was collapsing the batch dimension in my case because the base action was rank 0 (i.e. shape(action) = [B, ]).

I'm changing the PR to swap the local loss function to `squared_difference` if the single continuous action is rank 0. But keep the proper MSE otherwise.